File: bot.js
const puppeteer = require('puppeteer-extra');
const StealthPlugin = require('puppeteer-extra-plugin-stealth');
puppeteer.use(StealthPlugin());

const { PuppeteerScreenRecorder } = require('puppeteer-screen-recorder');
const cloudinary = require('cloudinary').v2;
const admin = require('firebase-admin');
const path = require('path');
const fs = require('fs');

cloudinary.config({
  cloud_name: process.env.CLOUDINARY_NAME,
  api_key: process.env.CLOUDINARY_API_KEY,
  api_secret: process.env.CLOUDINARY_API_SECRET,
});

if (!admin.apps.length) {
  admin.initializeApp({
    credential: admin.credential.cert(JSON.parse(process.env.FIREBASE_KEY))
  });
}
const db = admin.firestore();

async function runBot() {
  console.log("🚀 GHOST_ENGINE: V15 - Cookie Injection & High-Def Recording...");
  
  const browser = await puppeteer.launch({ 
    headless: "new", // GitHub Actions ke liye 'new' best h
    args: [
      '--no-sandbox', 
      '--disable-setuid-sandbox',
      '--disable-blink-features=AutomationControlled',
      '--window-size=1280,720'
    ] 
  });
  
  const page = await browser.newPage();
  
  // Mobile Viewport Force (iPhone 13 Pro layout)
  await page.setViewport({ width: 390, height: 844 }); 
  await page.setUserAgent('Mozilla/5.0 (iPhone; CPU iPhone OS 16_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.0 Mobile/15E148 Safari/604.1');

  const videoPath = path.join(__dirname, 'ghost_action.mp4');
  const recorder = new PuppeteerScreenRecorder(page, {
    followNewTab: true,
    fps: 25, // Strong video ke liye FPS badha diya
    videoFrame: { width: 390, height: 844 },
    aspectRatio: '9:16',
  });

  try {
    // 1. Start Recording BEFORE anything else
    await recorder.start(videoPath);
    console.log("⏺️ Recording Started...");

    // --- 🍪 STEP: COOKIE INJECTION (Tumhare App wala Data) ---
    if (process.env.INSTA_COOKIES) {
        console.log("🍪 Injecting Secret Cookies from App...");
        const cookies = JSON.parse(process.env.INSTA_COOKIES);
        await page.setCookie(...cookies);
    }

    // 2. Visit Instagram
    console.log("🌐 Opening Instagram...");
    await page.goto('https://www.instagram.com/', { waitUntil: 'networkidle2', timeout: 60000 });
    await new Promise(r => setTimeout(r, 5000)); // Render hone ka wait karo

    // 3. Verification: Kya hum login ho chuke hain?
    let isLoggedIn = await page.evaluate(() => {
        return !!document.querySelector('svg[aria-label="Home"]') || !!document.querySelector('nav');
    });

    // --- 🛠️ FALLBACK: Agar Cookie kaam nahi ki toh Manual Login ---
    if (!isLoggedIn && process.env.INSTA_PASS) {
        console.log("⚠️ Cookies failed or expired. Trying Manual Login...");
        await page.goto('https://www.instagram.com/accounts/login/', { waitUntil: 'networkidle2' });
        
        const targetUser = process.env.INSTA_USER || "ayush_raj6888";
        const targetPass = process.env.INSTA_PASS;

        await page.waitForSelector('input[name="username"]', { visible: true });
        await page.type('input[name="username"]', targetUser, { delay: 100 });
        await page.type('input[name="password"]', targetPass, { delay: 100 });
        
        await page.click('button[type="submit"]');
        console.log("⏳ Waiting for manual login redirect...");
        await new Promise(r => setTimeout(r, 15000));
        
        isLoggedIn = await page.evaluate(() => !!document.querySelector('nav'));
    }

    if (!isLoggedIn) {
      console.log("❌ LOGIN FAIL: Check the video later.");
    } else {
      console.log("✅ LOGIN SUCCESS! Bot is in.");

      const targets = ["_anshu_2101", "_cool_butterfly_.6284", "dee_pu3477", "ritu_singh785903"];
      
      for (const user of targets) {
        console.log(`📡 Scanning: @${user}`);
        await page.goto(`https://www.instagram.com/${user}/`, { waitUntil: 'networkidle2' });
        await new Promise(r => setTimeout(r, 5000)); // Strong video ke liye wait
        
        // Screenshot for debug
        console.log(`📸 Capturing @${user}`);
        
        const media = await page.evaluate(() => {
          const results = [];
          const items = document.querySelectorAll('img[srcset], article img, video, div._aagv img');
          items.forEach(el => {
            const src = el.src || el.srcset?.split(' ')[0];
            if (src && src.includes('cdninstagram.com')) results.push(src);
          });
          return [...new Set(results)];
        });

        console.log(`📊 Found ${media.length} items for @${user}`);
        for (const mUrl of media) {
            await safeUpload(mUrl, user, mUrl.includes('.mp4') ? 'videos' : 'posts');
        }
      }
    }

  } catch (error) {
    console.error("❌ CRITICAL ERROR:", error.message);
  } finally {
    console.log("⏹️ Mission Complete. Stopping Recorder...");
    await recorder.stop();
    await browser.close();
    
    // Video upload to Cloudinary for proof
    if (fs.existsSync(videoPath)) {
      console.log("☁️ Uploading Evidence Video...");
      await cloudinary.uploader.upload(videoPath, { 
        resource_type: "video", 
        folder: "ghost/sessions",
        public_id: `session_${Date.now()}`
      }).then(res => console.log("🎬 Video Link:", res.secure_url))
      .catch(e => console.log("Video Upload Error:", e.message));
    }
  }
}

async function safeUpload(url, username, category) {
  try {
    const mediaId = url.split('?')[0].split('/').pop().substring(0, 45); 
    const docId = `${username}_${mediaId}`;
    const docRef = db.collection("archives").doc(docId);
    
    const doc = await docRef.get();
    if (doc.exists) return; 

    const upload = await cloudinary.uploader.upload(url, { 
        folder: `insta_vault/${username}/${category}`, 
        resource_type: "auto" 
    });

    await docRef.set({ 
        owner: username, 
        url: upload.secure_url, 
        type: category, 
        time: admin.firestore.FieldValue.serverTimestamp() 
    });
    console.log(`      ✅ Saved to Vault: ${category}`);
  } catch (e) {
    console.log("      ❌ Upload Skip");
  }
}

runBot();

File: bot.py
import os
import json
import asyncio
import cloudinary
import cloudinary.uploader
import firebase_admin
from firebase_admin import credentials, firestore
import google.generativeai as genai
from playwright.async_api import async_playwright
from playwright_stealth import stealth_async
import requests

# ⚙️ CONFIG (Aapke Variables)
cloudinary.config(
    cloud_name=os.environ.get('CLOUDINARY_NAME'),
    api_key=os.environ.get('CLOUDINARY_API_KEY'),
    api_secret=os.environ.get('CLOUDINARY_API_SECRET')
)

genai.configure(api_key=os.environ.get('GEMINI_API_KEY'))

if not firebase_admin._apps:
    cred_json = json.loads(os.environ.get('FIREBASE_KEY'))
    firebase_admin.initialize_app(credentials.Certificate(cred_json))
db = firestore.client()

captured_media_urls = {} # URL store karne ke liye dictionary

# --- Gemini AI Analysis ---
async def analyze_with_gemini(image_url):
    try:
        model = genai.GenerativeModel('gemini-1.5-flash-latest')
        prompt = "Identify Area and Describe: [Area] - [Action]"
        response = requests.get(image_url)
        if response.status_code == 200:
            result = model.generate_content([
                prompt,
                {"mime_type": "image/jpeg", "data": response.content}
            ])
            return result.text
    except Exception as e:
        print(f"      ⚠️ AI Error: {e}")
    return "AI analysis unavailable."

# --- Firebase & Cloudinary Sync ---
async def safe_sync(url, username, category, is_video):
    try:
        media_id = url.split('?')[0].split('/')[-1][:45]
        doc_id = f"V28_PY_{username}_{media_id}"
        doc_ref = db.collection("archives").document(doc_id)

        if doc_ref.get().exists:
            return False, None

        print(f"      📤 Uploading to Cloudinary...")
        upload = cloudinary.uploader.upload(
            url, 
            folder=f"insta_vault_v28/{username}/{category}",
            resource_type="video" if is_video else "image"
        )

        doc_ref.set({
            "owner": username,
            "url": upload['secure_url'],
            "type": category,
            "is_video": is_video,
            "time": firestore.SERVER_TIMESTAMP,
            "version": "V28_PLAYWRIGHT_PYTHON"
        })
        return True, doc_id
    except Exception as e:
        print(f"      ❌ Sync Error: {e}")
        return False, None

# --- Network Interception Logic ---
async def handle_request(request):
    url = request.url
    if "instagram.com" in url and ("video" in url or ".mp4" in url or ".jpg" in url):
        if "profile_pic" not in url and "data:image" not in url:
            is_video = "video" in url or ".mp4" in url
            captured_media_urls[url] = {"type": "video" if is_video else "image"}
            # print(f"         🌐 Network captured: {url[:60]}...")

async def run_bot():
    async with async_playwright() as p:
        print("🚀 GHOST_ENGINE: V28 (Python) - NETWORK FALLBACK ACTIVATED...")
        browser = await p.chromium.launch(headless=True, args=["--no-sandbox"])
        context = await browser.new_context(viewport={'width': 1080, 'height': 1920})
        
        # Apply Stealth
        page = await context.new_page()
        await stealth_async(page)

        # 🍪 Apply Cookies
        if os.environ.get('INSTA_COOKIES'):
            cookies = json.loads(os.environ.get('INSTA_COOKIES'))
            await context.add_cookies(cookies)

        # Enable Network Interception
        page.on("request", handle_request)

        targets = ["_anshu_2101", "_cool_butterfly_.6284", "dee_pu3477", "ritu_singh785903"]

        for user in targets:
            print(f"\n🕵️ Target Locked: @{user}")
            try:
                # Stories page par jana
                await page.goto(f"https://www.instagram.com/stories/{user}/", wait_until="networkidle")
                await asyncio.sleep(6) # Initial load wait

                for slide in range(10): # Max 10 stories per user
                    if "/stories/" not in page.url: break
                    
                    print(f"   🎯 SLIDE {slide + 1}: Capturing...")
                    
                    # SCREEN SMASH: Overlays hatane ke liye click
                    await page.mouse.click(540, 960)
                    await asyncio.sleep(1)

                    # Capture and Sync
                    if captured_media_urls:
                        latest_url = list(captured_media_urls.keys())[-1]
                        media_data = captured_media_urls[latest_url]
                        
                        is_video = media_data['type'] == 'video'
                        is_new, doc_id = await safe_sync(latest_url, user, "stories", is_video)
                        
                        if is_new and not is_video:
                            desc = await analyze_with_gemini(latest_url)
                            db.collection("archives").document(doc_id).update({"ai_report": desc})
                    
                    # Next Story
                    await page.keyboard.press("ArrowRight")
                    await asyncio.sleep(3)

            except Exception as e:
                print(f"❌ Error scanning @{user}: {e}")

        await browser.close()
        print("⏹️ Mission Finished.")

if __name__ == "__main__":
    asyncio.run(run_bot())
  
